clazy

https://github.com/KDE/clazy

Best practices checker for Qt (C++) with fixits.

I've started on this (will push a PR soon), but can't get it working.

It seems like `clazy-standalone` is actually trying to compile things (or expects it to somehow have been compiled). Our test case fails with:

```
       Error while trying to load a compilation database:
       Could not auto-detect compilation database for file "./clazy-test-1.cpp"
       No compilation database found in /code or any parent directory
       fixed-compilation-database: Error while opening fixed database: No such file or directory
       json-compilation-database: Error while opening JSON database: No such file or directory
       Running without flags.
       /code/clazy-test-1.cpp:1:10: fatal error: 'QString' file not found
       #include <QString>
                ^~~~~~~~~
       1 error generated.
       Error while processing /code/clazy-test-1.cpp.
```

I'm not sure what would be required to get it working.

--- a/_tools/restylers/restylers.cabal
+++ b/_tools/restylers/restylers.cabal
@@ -4,7 +4,7 @@ cabal-version: 1.12
 --
 -- see: https://github.com/sol/hpack
 --
--- hash: c07856477d78bb71342f7c73e16eae8ad05c4c3985ae42c271dbfa0252bc4f4c
+-- hash: e3541df039aa6277437eb1f4ce9c39e96d53b63fba9fad21990854be74cbb8f0
 
 name:           restylers
 version:        0.0.0.0
@@ -29,6 +29,7 @@ library
       Restylers.Override
       Restylers.Prelude
       Restylers.Test
+      Restylers.Test.Output
       Restylers.Version
   other-modules:
       Paths_restylers

--- a/_tools/restylers/src/Restylers/Info/Test.hs
+++ b/_tools/restylers/src/Restylers/Info/Test.hs
@@ -10,8 +10,8 @@
 -- Portability : POSIX
 module Restylers.Info.Test
   ( Test (..)
-  , writeTestFiles
-  , testFilePath
+  , writeTestFile
+  , writeTestSupportFile
   , testDescription
   )
 where
@@ -22,7 +22,7 @@ import Data.Aeson
 import Data.List.NonEmpty qualified as NE
 import Data.Text qualified as T
 import Data.Text.IO qualified as T
-import Restylers.Info.Test.Support (Support)
+import Restylers.Info.Test.Support (Support (..))
 import Restylers.Name
 import System.FilePath (takeExtension)
 
@@ -36,20 +36,22 @@ data Test = Test
   deriving stock (Eq, Show, Generic)
   deriving anyclass (FromJSON, ToJSON)
 
-writeTestFiles
-  :: ( MonadIO m
-     , MonadLogger m
-     )
+writeTestFile
+  :: MonadIO m
   => Int
   -> RestylerName
   -> [Text]
   -> Test
   -> m FilePath
-writeTestFiles number name include test@Test {contents} = do
-  logInfo $ "CREATE" :# ["path" .= path]
+writeTestFile number name include test@Test {contents} = do
   path <$ liftIO (T.writeFile path contents)
  where
   path = testFilePath number name include test
+
+writeTestSupportFile :: MonadIO m => Test -> m ()
+writeTestSupportFile Test {support} = do
+  for_ support $ \Support {path, contents} ->
+    liftIO $ T.writeFile path contents
 
 testFilePath :: Int -> RestylerName -> [Text] -> Test -> FilePath
 testFilePath number name include Test {extension} =

--- a/_tools/restylers/src/Restylers/Info/Test.hs
+++ b/_tools/restylers/src/Restylers/Info/Test.hs
@@ -32,6 +32,8 @@ data Test = Test
   , contents :: Text
   , restyled :: Text
   , support :: Maybe Support
+  , command :: Maybe [Text]
+  , arguments :: Maybe [Text]
   }
   deriving stock (Eq, Show, Generic)
   deriving anyclass (FromJSON, ToJSON)

--- a/_tools/restylers/src/Restylers/Info/Test/Support.hs
+++ b/_tools/restylers/src/Restylers/Info/Test/Support.hs
@@ -1,5 +1,3 @@
-{-# LANGUAGE NamedFieldPuns #-}
-
 -- |
 --
 -- Module      : Restylers.Info.Test.Support
@@ -10,16 +8,12 @@
 -- Portability : POSIX
 module Restylers.Info.Test.Support
   ( Support (..)
-  , withSupportFile
   )
 where
 
 import Restylers.Prelude
 
 import Data.Aeson
-import Data.Text.IO qualified as T
-import UnliftIO.Directory (removeFile)
-import UnliftIO.Exception (finally)
 
 data Support = Support
   { path :: FilePath
@@ -27,12 +21,3 @@ data Support = Support
   }
   deriving stock (Eq, Show, Generic)
   deriving anyclass (FromJSON, ToJSON)
-
-withSupportFile
-  :: MonadUnliftIO m
-  => Support
-  -> m a
-  -> m a
-withSupportFile Support {path, contents} f = do
-  liftIO $ T.writeFile path contents
-  f `finally` removeFile path

--- a/_tools/restylers/src/Restylers/Test.hs
+++ b/_tools/restylers/src/Restylers/Test.hs
@@ -14,26 +14,24 @@ import Restylers.Prelude
 
 import Blammo.Logging.Logger (flushLogger)
 import Data.Aeson (ToJSON, object)
-import Data.ByteString.Lazy.Char8 qualified as BSL8
 import Data.Text.IO qualified as T
 import Data.Yaml qualified as Yaml
 import Restylers.Env
 import Restylers.Info.Metadata qualified as Metadata
 import Restylers.Info.Test
   ( testDescription
-  , testFilePath
-  , writeTestFiles
+  , writeTestFile
+  , writeTestSupportFile
   )
 import Restylers.Info.Test qualified as Test
-import Restylers.Info.Test.Support (withSupportFile)
 import Restylers.Manifest qualified as Manifest
 import Restylers.Name (RestylerName (..))
+import Restylers.Test.Output
 import System.Environment (withArgs)
-import System.FilePath (takeBaseName, (</>))
 import System.Process.Typed
 import Test.Hspec
 import UnliftIO.Directory
-import UnliftIO.Temporary (withTempDirectory)
+import UnliftIO.Temporary (withSystemTempDirectory)
 
 data CRestyler = CRestyler
   { name :: RestylerName
@@ -53,80 +51,73 @@ testRestylers
   -> [String]
   -> m ()
 testRestylers pull restylers hspecArgs = do
-  cwd <- getCurrentDirectory
   rts <- (.testShow) <$> getRestylersEnv
 
-  withTempDirectory cwd "restylers-test" $ \tmp ->
-    withCurrentDirectory tmp $ do
-      crestylers <- for restylers $ \restyler -> do
-        files <- for (restylerTests restyler) $ \(number, test) -> do
-          writeTestFiles number restyler.name restyler.include test
-        pure $ CRestyler restyler.name files
+  flushLogger -- before hspec makes its own output
+  liftIO $ do
+    withArgs hspecArgs $ hspec $ do
+      for_ restylers $ \restyler -> do
+        describe (unpack $ restyler.name.unwrap) $ do
+          for_ (restylerTests restyler) $ \(number, test) -> do
+            it (testDescription number test) $ do
+              inTempDirectory $ do
+                writeYaml testManifest restylers
+                writeTestSupportFile test
 
-      writeYaml testManifest restylers
-      writeYaml ".restyled.yaml"
-        $ object
-          [ "restylers_version" .= ("testing" :: Text)
-          , "restylers" .= crestylers
-          ]
+                file <- writeTestFile number restyler.name restyler.include test
 
-      let code = cwd </> takeBaseName tmp
+                writeYaml ".restyled.yaml"
+                  $ object
+                    [ "restylers_version" .= ("testing" :: Text)
+                    , "restylers"
+                        .= [ restyler
+                              { Manifest.enabled = True
+                              , Manifest.include = [pack file]
+                              }
+                           ]
+                    ]
 
-      flushLogger -- before hspec makes its own output
-      liftIO $ do
-        withArgs hspecArgs $ hspec $ do
-          for_ restylers $ \restyler -> do
-            describe (unpack $ restyler.name.unwrap) $ do
-              for_ (restylerTests restyler) $ \(number, test) -> do
-                it (testDescription number test) $ do
-                  (ec, out, err) <-
-                    maybe id withSupportFile test.support
-                      $ readProcess
-                      $ proc "restyle"
-                      $ concat
-                        [ ["--debug"]
-                        , ["--color", "always"]
-                        , ["--host-directory", code]
-                        , ["--manifest", testManifest]
-                        , ["--no-commit"]
-                        , ["--no-clean"]
-                        , ["--no-pull" | not pull]
-                        , ["--restyler-memory", "512m"]
-                        , ["."]
-                        ]
+                (ec, out, err) <-
+                  readProcess
+                    $ proc "restyle"
+                    $ concat
+                      [ ["--debug"]
+                      , ["--color", "always"]
+                      , ["--manifest", testManifest]
+                      , ["--no-commit"]
+                      , ["--no-clean"]
+                      , ["--no-pull" | not pull]
+                      , ["--restyler-memory", "512m"]
+                      , ["."]
+                      ]
 
-                  when (ec /= ExitSuccess) $ do
-                    throwString
-                      $ unlines
-                        [ "Restyler " <> show ec
-                        , "stdout:"
-                        , BSL8.unpack out
-                        , ""
-                        , "stderr:"
-                        , BSL8.unpack err
-                        , ""
-                        ]
+                when (ec /= ExitSuccess)
+                  $ expectationFailure
+                  $ "Restyler "
+                  <> show ec
+                  <> "\n"
+                  <> frameOutput out err
 
-                  restyled <- T.readFile $ testFilePath number restyler.name restyler.include test
+                restyled <- T.readFile file
 
-                  when (restyled == test.contents) $ do
-                    expectationFailure
-                      $ unlines
-                        [ "Restyler made no changes to the given files. For debugging, output was:"
-                        , "stdout:"
-                        , BSL8.unpack out
-                        , ""
-                        , "stderr:"
-                        , BSL8.unpack err
-                        , ""
-                        ]
+                when (restyled == test.contents)
+                  $ expectationFailure
+                  $ "Restyler made no changes to "
+                  <> file
+                  <> "\n"
+                  <> frameOutput out err
 
-                  if rts
-                    then show restyled `shouldBe` show test.restyled
-                    else restyled `shouldBe` test.restyled
+                if rts
+                  then show restyled `shouldBe` show test.restyled
+                  else restyled `shouldBe` test.restyled
 
 restylerTests :: Manifest.Restyler -> [(Int, Test.Test)]
 restylerTests r = zip [1 ..] r.metadata.tests
+
+inTempDirectory :: MonadUnliftIO m => m a -> m a
+inTempDirectory f = do
+  withSystemTempDirectory "restylers-test" $ \tmp ->
+    withCurrentDirectory tmp f
 
 writeYaml :: (MonadIO m, ToJSON a) => FilePath -> a -> m ()
 writeYaml path =

--- a/_tools/restylers/src/Restylers/Test.hs
+++ b/_tools/restylers/src/Restylers/Test.hs
@@ -73,6 +73,8 @@ testRestylers pull restylers hspecArgs = do
                         .= [ restyler
                               { Manifest.enabled = True
                               , Manifest.include = [pack file]
+                              , Manifest.command = fromMaybe restyler.command test.command
+                              , Manifest.arguments = fromMaybe restyler.arguments test.arguments
                               }
                            ]
                     ]

--- a/_tools/restylers/src/Restylers/Test.hs
+++ b/_tools/restylers/src/Restylers/Test.hs
@@ -109,6 +109,18 @@ testRestylers pull restylers hspecArgs = do
 
                   restyled <- T.readFile $ testFilePath number restyler.name restyler.include test
 
+                  when (restyled == test.contents) $ do
+                    expectationFailure
+                      $ unlines
+                        [ "Restyler made no changes to the given files. For debugging, output was:"
+                        , "stdout:"
+                        , BSL8.unpack out
+                        , ""
+                        , "stderr:"
+                        , BSL8.unpack err
+                        , ""
+                        ]
+
                   if rts
                     then show restyled `shouldBe` show test.restyled
                     else restyled `shouldBe` test.restyled

--- a/_tools/restylers/src/Restylers/Test/Output.hs
+++ b/_tools/restylers/src/Restylers/Test/Output.hs
@@ -1,0 +1,40 @@
+module Restylers.Test.Output
+  ( frameOutput
+  ) where
+
+import Restylers.Prelude
+
+import Data.ByteString.Lazy qualified as BSL
+import Data.ByteString.Lazy.Char8 qualified as BSL8
+
+{-
+
+╭─┤ stdout ├───────
+│ some output here
+│ more output here
+╰─
+
+╭─┤ stderr ├───────
+│ some output here
+│ more output here
+╰─
+
+-}
+frameOutput :: BSL.ByteString -> BSL.ByteString -> String
+frameOutput stdout stderr =
+  unlines
+    $ map ("  " <>)
+    $ frameOutputLines "stdout" (lines $ BSL8.unpack stdout)
+    <> frameOutputLines "stderr" (lines $ BSL8.unpack stderr)
+
+frameOutputLines :: String -> [String] -> [String]
+frameOutputLines name = (headerLine :) . (<> [footerLine]) . map outLine
+ where
+  headerLine :: String
+  outLine :: String -> String
+  footerLine :: String
+
+  -- Grouped below types so we can see it
+  headerLine = "╭─┤ " <> name <> "├──────"
+  outLine st = "│ " <> st
+  footerLine = "╰─"

--- a/clazy/Dockerfile
+++ b/clazy/Dockerfile
@@ -1,0 +1,21 @@
+FROM ubuntu:22.04
+LABEL maintainer="Pat Brisbin pbrisbin@restyled.io"
+ENV DEBIAN_FRONTEND=noninteractive LANG=C.UTF-8 LC_ALL=C.UTF-8
+RUN \
+  apt-get update && \
+  apt-get -y --no-install-recommends install \
+    clang \
+    clang-tools \
+    clazy \
+    cmake \
+    make \
+    locales \
+    qt6-base-dev && \
+  locale-gen en_US.UTF-8 && \
+  rm -rf /var/lib/apt/lists/*
+ENV LANG=en_US.UTF-8
+COPY clazy-standlone-apply /usr/local/bin/clazy-standalone-apply
+RUN mkdir -p /code
+WORKDIR /code
+ENTRYPOINT []
+CMD ["clazy-standalone-apply", "--help"]

--- a/clazy/clazy-standlone-apply
+++ b/clazy/clazy-standlone-apply
@@ -1,0 +1,9 @@
+#!/usr/bin/env bash
+set -eou pipefail
+
+tmp=$(mktemp -d)
+trap 'rm -rf "$tmp"' EXIT
+
+cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=ON .
+clazy-standalone --export-fixes="$tmp"/fixes.yaml "$@"
+clang-apply-replacements "$tmp"

--- a/clazy/info.yaml
+++ b/clazy/info.yaml
@@ -1,0 +1,62 @@
+enabled: true
+name: clazy
+command:
+  - clazy-standalone-apply
+supports_arg_sep: false
+version_cmd: |
+  apt-cache policy clazy | sed '/^ *Installed: \(.*\)$/!d; s//v\1/'
+include:
+  - "**/*.cpp"
+documentation:
+  - https://github.com/KDE/clazy
+metadata:
+  languages:
+    - C++
+  tests:
+    - name: qstring-allocations
+      arguments: ["-checks=level2"]
+      support:
+        path: CMakeLists.txt
+        contents: |
+          cmake_minimum_required(VERSION 3.16)
+
+          project(helloworld VERSION 1.0.0 LANGUAGES CXX)
+
+          set(CMAKE_CXX_STANDARD 17)
+          set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+          set(CMAKE_AUTOMOC ON)
+
+          find_package(Qt6 REQUIRED COMPONENTS Core)
+
+          add_executable(helloworld
+              clazy-test-1.cpp
+          )
+
+          target_link_libraries(helloworld PRIVATE Qt6::Core)
+      contents: |
+        #include <iostream>
+        #include <QString>
+
+        int main() {
+          QString str = "hello";
+          QString result = str.left(1);
+
+          std::cout << "Original string: " << str.toStdString() << std::endl;
+          std::cout << "Substring from index 5 onwards: " << result.toStdString() << std::endl;
+
+          return 0;
+        }
+      restyled: |
+        #include <iostream>
+        #include <QString>
+
+        int main() {
+          QString str = QStringLiteral("hello");
+          QString result = str.left(1);
+
+          std::cout << "Original string: " << str.toStdString() << std::endl;
+          std::cout << "Substring from index 5 onwards: " << result.toStdString() << std::endl;
+
+          return 0;
+        }

